registry/auth/openshift/accessController.go

```diff
@@ -0,0 +1,148 @@
+)
+
+type accessController struct {
+	openshiftAddr       string
+	openshiftApiVersion string
+	client              *http.Client
+}
+
+type openshiftToken struct {
+	namespace   string
+	imageName   string
+	verb        string
+	bearerToken string
+}
+
+var _ auth.AccessController = &accessController{}
+
+func (ac *accessController) verifyOpenShiftToken(token openshiftToken) error {
+	url := fmt.Sprintf("https://%s/subjectAccessReviews?namespace=%s", ac.openshiftAddr, token.namespace)
+	sar := map[string]string{
+		"kind":         "SubjectAccessReview",
+		"apiVersion":   ac.openshiftApiVersion,
+		"verb":         token.verb,
+		"resource":     "imageRepositories",
+		"resourceName": token.imageName,
+	}
+
+	body, err := json.Marshal(sar)
+	if err != nil {
+		return fmt.Errorf("Error marshaling SubjectAccessReview: %s", sar)
+	}
+	req, err := http.NewRequest("POST", url, strings.NewReader(string(body)))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenShift request: %s", err)
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.bearerToken))
+
+	resp, err := ac.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Error querying OpenShift for SubjectAccessReview: %s", err)
+	}
+
+	defer resp.Body.Close()
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("Error reading OpenShift SubjectAccessReview response body: %s", err)
+	}
+
+	var accessResponse authorizationapi.SubjectAccessReviewResponse
+	err = json.Unmarshal(respBody, &accessResponse)
+	if err != nil {
+		return fmt.Errorf("Error parsing OpenShift SubjectAccessReviewResponse: %s", err)
+	}
+	if !accessResponse.Allowed {
+		return fmt.Errorf("OpenShift access denied: %s", accessResponse.Reason)
+	}
+	return nil
+}
+
+func (ac *accessController) Authorized(ctx context.Context, accessRecords ...auth.Access) (context.Context, error) {
+	req, err := ctxu.GetRequest(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(accessRecords) == 0 {
+		return ctx, nil
+	}
+	for _, access := range accessRecords {
+		//fmt.Printf("%s:%s:%s", access.Resource.Type, access.Resource.Name, access.Action)
+
+		if access.Resource.Type != "repository" {
+			continue
+		}
+		repoParts := strings.Split(access.Resource.Name, "/")
+		if len(repoParts) != 2 {
+			return nil, fmt.Errorf("Repository namespace not found for resource: %s", access.Resource.Name)
+		}
+		var token openshiftToken
+		token.namespace, token.imageName = repoParts[0], repoParts[1]
+
+		if access.Action == "push" {
+			token.verb = "create"
+		} else if access.Action == "pull" {
+			token.verb = "get"
+		} else {
+			return nil, fmt.Errorf("Unkown action: %s", access.Action)
+		}
+
+		authorization := req.Header.Get("Authorization")
+		if len(authorization) == 0 {
+			return nil, fmt.Errorf(`"Authorization" header not set`)
+		}
+		authParts := strings.Split(authorization, " ")
+		if len(authParts) != 2 || strings.ToLower(authParts[0]) != "bearer" {
```

Comment:
We're not going to be able to support "Authorization: bearer" in the short term, because the Docker daemon implements a very specific workflow for bearer tokens (https://github.com/docker/distribution/blob/master/doc/spec/auth/token.md).

The flow will need to be:

1. acquire oauth token from OpenShift
1. `docker login`, and when it asks for password, specify the oauth token received in step 1
1. when you `docker pull` or `docker push`, docker will send the username and "password" (i.e. token) in an `Authorization: basic` header